File: app/src/main/java/com/aurora/store/sheet/DownloadMenuSheet.java
/*
 * Aurora Store
 * Copyright (C) 2019, Rahul Kumar Patel <whyorean@gmail.com>
 *
 * Aurora Store is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Aurora Store is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Aurora Store.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 */

package com.aurora.store.sheet;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.aurora.store.R;
import com.aurora.store.download.DownloadManager;
import com.aurora.store.util.Util;
import com.google.android.material.navigation.NavigationView;
import com.tonyodev.fetch2.Fetch;
import com.tonyodev.fetch2.Status;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DownloadMenuSheet extends BaseBottomSheet {

    public static final String TAG = "DOWNLOAD_MENU_SHEET";

    public static final String DOWNLOAD_ID = "DOWNLOAD_ID";
    public static final String DOWNLOAD_STATUS = "DOWNLOAD_STATUS";
    public static final String DOWNLOAD_URL = "DOWNLOAD_URL";

    @BindView(R.id.navigation_view)
    NavigationView navigationView;

    private Fetch fetch;

    private int id;
    private int status;
    private String url;

    public DownloadMenuSheet() {
    }

    @Override
    public View onCreateContentView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.sheet_download_menu, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onContentViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getArguments() != null) {
            Bundle bundle = getArguments();
            id = bundle.getInt(DOWNLOAD_ID);
            status = bundle.getInt(DOWNLOAD_STATUS);
            url = bundle.getString(DOWNLOAD_URL);
            fetch = DownloadManager.getFetchInstance(requireContext());
            setupNavigation();
        } else {
            dismissAllowingStateLoss();
        }
    }

    private void setupNavigation() {
        if (status == Status.PAUSED.getValue()
                || status == Status.COMPLETED.getValue()
                || status == Status.CANCELLED.getValue()) {
            navigationView.getMenu().findItem(R.id.action_pause).setVisible(false);
        }

        if (status == Status.DOWNLOADING.getValue()
                || status == Status.COMPLETED.getValue()
                || status == Status.QUEUED.getValue()) {
            navigationView.getMenu().findItem(R.id.action_resume).setVisible(false);
        }

        if (status == Status.COMPLETED.getValue()
                || status == Status.CANCELLED.getValue()) {
            navigationView.getMenu().findItem(R.id.action_cancel).setVisible(false);
        }

        navigationView.setNavigationItemSelectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.action_copy:
                    Util.copyToClipBoard(requireContext(), url);
                    Toast.makeText(requireContext(), requireContext().getString(R.string.action_copied), Toast.LENGTH_LONG).show();
                    break;
                case R.id.action_pause:
                    fetch.pause(id);
                    break;
                case R.id.action_resume:
                    if (status == Status.FAILED.getValue()
                            || status == Status.CANCELLED.getValue())
                        fetch.retry(id);
                    else
                        fetch.resume(id);
                    break;
                case R.id.action_cancel:
                    fetch.cancel(id);
                    break;
                case R.id.action_clear:
                    fetch.delete(id);
                    break;
            }
            dismissAllowingStateLoss();
            return false;
        });
    }
}

File: app/src/main/java/com/aurora/store/view/ui/sheets/DownloadMenuSheet.kt
/*
 * Aurora Store
 *  Copyright (C) 2021, Rahul Kumar Patel <whyorean@gmail.com>
 *
 *  Aurora Store is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  Aurora Store is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Aurora Store.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.aurora.store.view.ui.sheets

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.viewModels
import androidx.lifecycle.findViewTreeLifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.navArgs
import com.aurora.extensions.copyToClipBoard
import com.aurora.extensions.toast
import com.aurora.store.R
import com.aurora.store.data.installer.AppInstaller
import com.aurora.store.data.model.DownloadStatus
import com.aurora.store.databinding.SheetDownloadMenuBinding
import com.aurora.store.util.DownloadWorkerUtil
import com.aurora.store.util.PathUtil
import com.aurora.store.viewmodel.sheets.SheetsViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class DownloadMenuSheet : BaseDialogSheet<SheetDownloadMenuBinding>() {

    private val TAG = DownloadMenuSheet::class.java.simpleName
    private val viewModel: SheetsViewModel by viewModels()
    private val args: DownloadMenuSheetArgs by navArgs()
    private val playStoreURL = "https://play.google.com/store/apps/details?id="

    private val exportMimeType = "application/zip"

    private val requestDocumentCreation =
        registerForActivityResult(ActivityResultContracts.CreateDocument(exportMimeType)) {
            if (it != null) {
                viewModel.copyDownloadedApp(requireContext(), args.download, it)
            } else {
                toast(R.string.failed_apk_export)
            }
            dismissAllowingStateLoss()
        }

    @Inject
    lateinit var downloadWorkerUtil: DownloadWorkerUtil

    @Inject
    lateinit var appInstaller: AppInstaller

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        with(binding.navigationView) {
            val downloadCompleted = args.download.downloadStatus == DownloadStatus.COMPLETED
            val downloadDir = PathUtil.getAppDownloadDir(
                requireContext(),
                args.download.packageName,
                args.download.versionCode
            )

            menu.findItem(R.id.action_cancel).isVisible = !args.download.isFinished
            menu.findItem(R.id.action_clear).isVisible = args.download.isFinished
            menu.findItem(R.id.action_install).isVisible = downloadCompleted
            menu.findItem(R.id.action_local).isVisible =
                downloadCompleted && downloadDir.listFiles() != null

            setNavigationItemSelectedListener { item ->
                when (item.itemId) {
                    R.id.action_install -> {
                        install()
                        dismissAllowingStateLoss()
                    }

                    R.id.action_copy -> {
                        requireContext().copyToClipBoard(
                            "${playStoreURL}${args.download.packageName}"
                        )
                        requireContext().toast(requireContext().getString(R.string.toast_clipboard_copied))
                        dismissAllowingStateLoss()
                    }

                    R.id.action_cancel -> {
                        findViewTreeLifecycleOwner()?.lifecycleScope?.launch(NonCancellable) {
                            downloadWorkerUtil.cancelDownload(args.download.packageName)
                        }
                        dismissAllowingStateLoss()
                    }

                    R.id.action_clear -> {
                        findViewTreeLifecycleOwner()?.lifecycleScope?.launch(NonCancellable) {
                            downloadWorkerUtil.clearDownload(
                                args.download.packageName,
                                args.download.versionCode
                            )
                        }
                        dismissAllowingStateLoss()
                    }

                    R.id.action_local -> {
                        requestDocumentCreation.launch("${args.download.packageName}.zip")
                    }
                }
                false
            }
        }
    }

    private fun install() {
        try {
            appInstaller.getPreferredInstaller().install(args.download)
        } catch (exception: Exception) {
            Log.e(TAG, "Failed to install ${args.download.packageName}", exception)
            if (exception is NullPointerException) {
                requireContext().toast(R.string.installer_status_failure_invalid)
            }
        }
    }
}
